Add non-catching `retrieve` methods

hamorillo opened the pull request:
Closes #212 

### Description

In a previous commit, we added the retrieveCatching methods, which prevent them from throwing exceptions. Now, we are adding the throwing exception version of them based on [this discussion](https://github.com/Automattic/Gravatar-SDK-Android/pull/207#discussion_r1664026712).

### Testing Steps
- CI and code review should be enough

hamorillo committed:
Add non-catching retrieve methods

In a previous commit, we added the retrieveCatching methods, which prevent them from throwing exceptions. Now, we are adding the throwing exception version of them.

AdamGrzybkowski commented on gravatar/src/main/java/com/gravatar/services/ProfileService.kt:
```diff
@@ -122,35 +123,81 @@ public class ProfileService(okHttpClient: OkHttpClient? = null) {
+     * @param email The email address to fetch the profile for
+     * @return The fetched profiles
+     */
+    public suspend fun retrieve(email: Email): Result<Profile, ErrorType> {
```
```suggestion
    public suspend fun retrieve(email: Email): Profile {
```

I think that we should return only when the operation is successful thus we don't need the `Result` return type. All errors should be thrown. Otherwise, users will have to implement two error handling for this method. Thoughts on this?

hamorillo replied:
> users will have to implement two error handling for this method

Yeah, that's a good point. We'll need to figure out what exceptions to throw and whether we need to declare our own exceptions for each of the current ErrorTypes. Anyway, I think it's a great suggestion. I'll iterate over it. Thanks @AdamGrzybkowski .

hamorillo replied:
I've pushed some changes. It should be ready for another round. 🙏🏼

hamorillo committed:
Non-catching methods throw exceptions for all kind of errors

For handling errors, we were returning a Result<Profile, ErrorType> and throwing exceptions at the same time. That will force developers to handle errors in two different ways, looking at the ErrorType and catching possible exceptions.

With these changes, we throw an exception for all errors so developers will receive a Profile or an exception.

AdamGrzybkowski commented on gravatar/src/main/java/com/gravatar/services/ProfileService.kt:
```diff
@@ -99,58 +99,99 @@ public class ProfileService(okHttpClient: OkHttpClient? = null) {
+    public suspend fun retrieve(hashOrUsername: String): Profile = withContext(GravatarSdkDI.dispatcherIO) {
+        val response = service.getProfileById(hashOrUsername)
+        if (response.isSuccessful) {
+            response.body() ?: error("Response body is null")
```
Quick question: 

Does `service.getProfileById(hashOrUsername)` always return a value even if the hash is not matched?  It will return an empty profile, right?

hamorillo replied:
The endpoint returns a `404` if the hash is not matched. We match that `404` to an empty state

<img width="1261" alt="image" src="https://github.com/user-attachments/assets/63ba83ed-4ab6-4b8f-af2e-4280a4ed483a">

AdamGrzybkowski replied:
>We match that 404 to an empty state

Do we match that in the SDK? I can only see that in the demo app. 


I think we shouldn't throw an error if it is expected not to get a profile. It's probably better to return null instead.

hamorillo replied:
Updated. Let me know what you think.

AdamGrzybkowski commented on gravatar/src/main/java/com/gravatar/services/ServiceUtils.kt:
```diff
@@ -3,6 +3,11 @@ package com.gravatar.services
+    replaceWith = ReplaceWith("com.gravatar.services.runCatchingRequest"),
+    level = DeprecationLevel.WARNING,
+)
 internal inline fun <T> runCatchingService(block: () -> Result<T, ErrorType>): Result<T, ErrorType> {
```
This is internal - do we have to deprecate it? Can't we remove it?

hamorillo replied:
I didn't remove it because is still in use [here](https://github.com/Automattic/Gravatar-SDK-Android/blob/1f754d9e778df05266e9810d432cc26695f44c07/gravatar/src/main/java/com/gravatar/services/ProfileService.kt#L34) so I think we can remove it when we remove the other deprecated methods.

AdamGrzybkowski commented on gravatar/src/test/java/com/gravatar/services/ProfileServiceTests.kt:
```diff
@@ -212,4 +203,99 @@ class ProfileServiceTests {
     }
+
+    @Test
+    fun `given an username when retrieving its profile and an exception is thrown then result is failure`() = runTest {
```
```suggestion
    fun `given a username when retrieving its profile and an exception is thrown then result is failure`() = runTest {
```

AdamGrzybkowski commented on gravatar/src/test/java/com/gravatar/services/ProfileServiceTests.kt:
```diff
@@ -212,4 +203,99 @@ class ProfileServiceTests {
+
+    // Throwing Exception Version of the methods
+    @Test(expected = TimeoutException::class)
+    fun `given an username when retrieving its profile and a timeout occurs then exception is thrown`() = runTest {
```
```suggestion
    fun `given a username when retrieving its profile and a timeout occurs then exception is thrown`() = runTest {
```

hamorillo committed:
Fix typo

Co-authored-by: Adam Grzybkowski <agrzybkowski@outlook.com>

hamorillo committed:
Return null on non-catching profile requests when the profile is not found.

AdamGrzybkowski reviewed (approved):
LGTM! 👍 Thanks for all the changes.
